### DISCUSSION TITLE
Show an explanation for why Actions workflows aren't run on PRs when there is a merge conflict

### DISCUSSION THREAD

**Discussion by danielcompton** [community/community#11265] [Actions]
I spent a chunk of time trying to understand why my Actions weren't running on a PR. Eventually I realised that it was because **Actions aren't triggered on PRs if there is a merge conflict**. This behaviour makes some amount of sense as the SHA that Actions uses is the merge commit, not the actual commit on the branch (#https://github.com/github/docs/issues/15302).

However, it would be great if there was an explanation in the PR UI explaining why. Currently everything is blank which made me think there was a problem with my workflow YAML file, but nothing showed up in the Actions list either.

Here’s what I see:

(The successful check in the UI is from an external system, not Actions)

![CleanShot 2022-02-10 at 09 51 14@2x](https://user-images.githubusercontent.com/811954/153287848-4c4a273d-e65d-4a71-b3eb-529494b00942.png)
![CleanShot 2022-02-10 at 09 48 28@2x](https://user-images.githubusercontent.com/811954/153287404-1bbd9a81-b770-453a-8622-42d3f8176b0f.png)

Related: https://github.community/t/run-actions-on-pull-requests-with-merge-conflicts/17104

**Comment by ethomson** [comment]
Hi @danielcompton - thanks for the feedback.  Your analysis is spot on - if we can't merge the PR then we can't build the merge result.  I'll talk to the team and see how we can make this clearer and we'll put some work on the backlog.  Thanks!

**Reply by chriscantu** [reply]
Any update on this issue?

**Reply by fru2zzy** [reply]
any solutions so far?

**Reply by mauritsdebruin** [reply]
plz fix

**Reply by rtk-willm** [reply]
this is still a confusing issue in 2024. Gitlab triggers CI runs on all commits so quite confusing for someone coming from that world trying to make sense as to why no checks are running. I spent some time triggering white space commits, empty commits and code commits in attempts to get checks working again, assuming there was some bug on GH end. 
A simple prompt in the UI explaining why no checks would save a lot of time.

**Reply by jjb** [reply]
> if we can't merge the PR then we can't build the merge result

i use and enjoy GHA extensively for many mission-critical things. NONE of my uses of it depend on the concept of "this must apply to the merge result." because of this behavior, i simply have to consider GHA to be nonfunctional until conflicts are resolved, which impedes many things for my team.

**Reply by asaphe** [reply]
any updates for this issue? this is very confusing. developers expect the checks to run even if there are conflicts. maybe we are doing something wrong but more than person have asked for this to be resolved.

**Comment by cjgraydon** [comment]
This is also a problem when a PR is not going to be merged directly, but will be resolved with a force push.

Scenario:

An internal branch is rebased on an upstream branch. We mirror an upstream branch to our internal repo, which we then rebase our internal changes on top weekly.

We open a PR with the rebased branch to run CI on it. If all goes well, we force push new version over old version of the internal branch.

If there is a merge conflict, however, it breaks the CI run.  Worse, if the conflict is due to something like an internal file being upstreamed and and removed on internal branch, OR is moved in the tree, it gets seen as a conflict, even though it is in the correct place when pulled in via the rebase. Note that this scenario resolves properly at the command line. But GitHub still sees it and breaks CI.

In other words, in the moved/deleted file scenario, there is literally no way to resolve a conflict at the command line that git doesn't see,  but GitHub still manages to spot it and fails to resolve in the same fashion that git does. Basically, a 3-way merge fails for GitHub that does not fail for git command line.

If there is a way to get around that, I'd love to know what it is.

**Comment by xkr47** [comment]
Imagine I am about to present a demo of the work I have done on my PR and we use the CI to deploy the latest version of the code to a publicly available website, to be used for the demo. Then, one hour before the demo I notice I need to make a small correction to my PR and have the CI deploy the new version of my PR, only to be demo-blocked by the CI because someone pushed a conflict into the main branch.

When you have branches with conflicts, from the PR's perspective it just means the conflicts need to be resolved before a merge/squash can be done. However merging/squashing may be be weeks away. Why should the CI block me from running lints, tests, deployments etc against the code until the conflicts are resolved? Why am I forced by the CI to stop all work and resolve every conflict that pops up before I can resume my work? It can also be a colossal waste of time to rebase every single time there is a conflict instead of resolving them all in one go later.

If the answer is "because it tries to merge with the target branch before running the tasks" then what about just not doing the merge in case of conflicts? It doesn't make the work of the workflows useless just because the merge cannot be performed this point in time. If there are some tasks that rely on the merge having happened, just make it optional like `run-despite-conflicts: true` or `run-without-merging-first: true` or even both in case the merging is not even desired.

**Reply by gaversano-irco** [reply]
Yep. Spot on. This is a bug.

**Reply by dietertao** [reply]
Is there any update on this topic? We’re facing a similar situation.

Quite often, we create a feature branch, for example due to feature development. In most cases, we don’t want to sync with main or develop immediately — we prefer to keep the original base. Once the work is completed, we then perform a rebase or merge from main/develop into our feature branch to bring it up to date.

In a pull request for the feature branch, we want the build and testing workflows to run. However, at the moment this isn’t possible because of merge conflicts, which prevent these workflows from triggering successfully.

**Comment by gaversano-irco** [comment]
Can't believe this is still an issue after years. Gitlab handles this superbly.
